doc/wol/Wake-on-LAN-HLD.md

```diff
@@ -0,0 +1,172 @@
+|--------------|---------------------------------------------------|
+| WoL          | **W**ake-**o**n-**L**AN                           |
+| gNOI         | **g**RPC **N**etwork **O**perations **I**nterface |
+| NIC          | **N**etwork **I**nterface **C**ontroller          |
+
+## Overview
+
+**Wake-on-LAN** (**WoL** or **WOL**) is an Ethernet or Token Ring computer networking standard that allows a computer to be turned on or awakened from sleep mode by a network message[^1]. This document describes the Wake-on-LAN feature design in SONiC.
+
+## Background
+
+Below diagram describes a common usage of WoL on SONiC switch, this HLD will focus on the green part:
+
+![WoL Background](./img/background.png)
+
+## Components
+
+### `wol` CLI script in sonic-utilities
+
+A `wol` script will be introduced in [sonic-utilities](https://github.com/sonic-net/sonic-utilities). The workflow of command line utility `wol` is:
+
+0. User login to the SONiC switch and enter the `wol` command.
+1. The `wol` script send magic packet to specific interface or VLAN.
+
+![Component Utilities](./img/component-utilities.png)
+
+### gNOI service
+
+A new gNOI service `SONiCWolService` will be implemented in sonic-gnmi container. The workflow is:
+
+0. User initialize a gNOI Client to communicate with gNOI server.
+1. The gNOI Client call the RPC function `SONiCWolService.Wol`.
+2. The gNOI server send a D-Bus request to sonic-host-service[^2].
+3. sonic-host-service call `wol` CLI to send the magic packet.
+
+![Component gNOI](./img/component-gnoi.png)
+
+## Magic Packet
+
+**Magic Packet** is a Ethernet frame with structure:
+
+* **Ethernet Frame Header**:
+  * **Destination MAC address**: broadcast MAC address (ff:ff:ff:ff:ff:ff) or target device's MAC address. [6 bytes]
+  * **Source MAC address**. [6 bytes]
+  * **EtherType**: `0x0842`. [2 bytes]
+* **Ethernet Frame Payload**:
+  * Six bytes of all `0xff`. [6 bytes]
+  * Sixteen repetitions of the target device's MAC address. [96 bytes]
+  * (Optional) A four or six byte password. [4 or 6 bytes]
+
+```
+Byte     |
+Offset   |0     |1     |2     |3     |4     |5     |
+---------+------+------+------+------+------+------+
+       0 |             Destination MAC             |
+         +------+------+------+------+------+------+        ETHERNET FRAME
+       6 |               Source MAC                |           HEADER
+         +------+------+---------------------------+
+      12 | 0x08 | 0x42 |   \      \      \      \
+         +------+------+------+------+------+------+------------------------------
+      14 | 0xFF | 0xFF | 0xFF | 0xFF | 0xFF | 0xFF |
+         +------+------+------+------+------+------+        ETHERNET FRAME
+      20 | Target MAC (repeat 16 times, 96 bytes)  |           PAYLOAD
+         +-----------------------------------------+
+     116 |    Password (optional, 4 or 6 bytes)    |
+         +-----------------------------------------+
```

Comment:
Wake on LAN UDP dest port -9 is not a requirement?

Reply:
The `wol` utility will construct and send an Ethernet frame. It doesn't require the packet is an IP packet.